### PR TITLE
Fix additional let/var init bugs

### PR DESCRIPTION
**Pull request by kzc**
Fix additional variable state change bugs including blockless `if` statement `var` declarations and `let`/`var` use before declaration within same function that do not cross function scopes.

This PR contains:
- [x] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?
- [x] yes (*bugfixes and features will not be merged without tests*)
- [ ] no

Breaking Changes?
- [ ] yes (*breaking changes will not be merged unless absolutely necessary*)
- [x] no

List any relevant issue numbers:

### Description

Fixes several outstanding var/let init issues mentioned in #4148 without any code pessimization other than for previously incorrect rollup output, and with minimal overhead.

### CODE REVIEW

**Review by lukastaegert — APPROVED** [review]
I took the liberty of replacing the removed example in the docs with one that is still failing, otherwise we can merge this for now.

**Comment by kzc** [comment]
Works for me.

File: docs/999-big-list-of-options.md [hunk 1]
```diff
@@ -1428,23 +1428,30 @@ Type: `boolean`<br>
 CLI: `--treeshake.correctVarValueBeforeDeclaration`/`--no-treeshake.correctVarValueBeforeDeclaration`<br>
 Default: `false`
 
-If a variable is assigned a value in its declaration and is never reassigned, Rollup sometimes assumes the value to be constant. This is not true if the variable is declared with `var`, however, as those variables can be accessed before their declaration where they will evaluate to `undefined`.
-Choosing `true` will make sure Rollup does not make (wrong) assumptions about the value of such variables. Note though that this can have a noticeable negative impact on tree-shaking results.
+If a variable is assigned a value in its declaration and is never reassigned, Rollup can in some adge case scenarios assume the value to be constant, see below. This is not true if the variable is declared with `var`, however, as those variables can be accessed before their declaration where they will evaluate to `undefined`.
```
Comment (kzc) [inline docs/999-big-list-of-options.md:1431]: There's a typo for edge. Also consider this phrasing for the first sentence:

"In some edge cases if a variable is accessed before its declaration assignment and is not reassigned then Rollup may incorrectly assume that variable is constant, as in the example below."

Comment (lukastaegert) [reply docs/999-big-list-of-options.md]: 👍

**Pull request merged**